urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZMA/Encoder.java

```diff
@@ -0,0 +1,1726 @@
+                            + posState]);
+        }
+        else
+        {
+            price = SevenZip.Compression.RangeCoder.Encoder
+                    .GetPrice1(_isRepG0[state]);
+            if (repIndex == 1)
+            {
+                price += SevenZip.Compression.RangeCoder.Encoder
+                        .GetPrice0(_isRepG1[state]);
+            }
+            else
+            {
+                price += SevenZip.Compression.RangeCoder.Encoder
+                        .GetPrice1(_isRepG1[state]);
+                price += SevenZip.Compression.RangeCoder.Encoder.GetPrice(
+                        _isRepG2[state], repIndex - 2);
+            }
+        }
+        return price;
+    }
+    
+    int GetRepPrice(int repIndex, int len, int state, int posState)
+    {
+        final int price = _repMatchLenEncoder.GetPrice(len - Base.kMatchMinLen,
+                posState);
+        return price + GetPureRepPrice(repIndex, state, posState);
+    }
+    
+    int GetPosLenPrice(int pos, int len, int posState)
+    {
+        int price;
+        final int lenToPosState = Base.GetLenToPosState(len);
+        if (pos < Base.kNumFullDistances)
+        {
+            price = _distancesPrices[lenToPosState * Base.kNumFullDistances
+                    + pos];
+        }
+        else
+        {
+            price = _posSlotPrices[(lenToPosState << Base.kNumPosSlotBits)
+                    + GetPosSlot2(pos)]
+                    + _alignPrices[pos & Base.kAlignMask];
+        }
+        return price + _lenEncoder.GetPrice(len - Base.kMatchMinLen, posState);
+    }
+    
+    int Backward(int cur)
+    {
+        _optimumEndIndex = cur;
+        int posMem = _optimum[cur].PosPrev;
+        int backMem = _optimum[cur].BackPrev;
+        do
+        {
+            if (_optimum[cur].Prev1IsChar)
+            {
+                _optimum[posMem].MakeAsChar();
+                _optimum[posMem].PosPrev = posMem - 1;
+                if (_optimum[cur].Prev2)
+                {
+                    _optimum[posMem - 1].Prev1IsChar = false;
+                    _optimum[posMem - 1].PosPrev = _optimum[cur].PosPrev2;
+                    _optimum[posMem - 1].BackPrev = _optimum[cur].BackPrev2;
+                }
+            }
+            final int posPrev = posMem;
+            final int backCur = backMem;
+            
+            backMem = _optimum[posPrev].BackPrev;
+            posMem = _optimum[posPrev].PosPrev;
+            
+            _optimum[posPrev].BackPrev = backCur;
+            _optimum[posPrev].PosPrev = cur;
+            cur = posPrev;
+        } while (cur > 0);
+        backRes = _optimum[0].BackPrev;
+        _optimumCurrentIndex = _optimum[0].PosPrev;
+        return _optimumCurrentIndex;
+    }
+    
+    int[] reps    = new int[Base.kNumRepDistances];
+    int[] repLens = new int[Base.kNumRepDistances];
+    int   backRes;
+    
+    int GetOptimum(int position) throws IOException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 342 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.